kore/src/Kore/Step/Simplification/SubstitutionSimplifier.hs

```diff
@@ -0,0 +1,147 @@
+{- |
+Copyright   : (c) Runtime Verification, 2018
+License     : NCSA
+
+ -}
+
+module Kore.Step.Simplification.SubstitutionSimplifier
+    ( SubstitutionSimplifier (..)
+    , simplification
+    , unification
+    ) where
+
+import Control.Error
+    ( maybeT
+    )
+import Data.Function
+    ( (&)
+    )
+import Data.Map.Strict
+    ( Map
+    )
+
+import Kore.Internal.OrPredicate
+    ( OrPredicate
+    )
+import qualified Kore.Internal.OrPredicate as OrPredicate
+import Kore.Internal.Predicate
+    ( Predicate
+    )
+import qualified Kore.Internal.Predicate as Predicate
+import Kore.Internal.TermLike
+    ( TermLike
+    )
+import qualified Kore.Predicate.Predicate as Syntax
+    ( Predicate
+    )
+import qualified Kore.Predicate.Predicate as Syntax.Predicate
+import Kore.Step.Simplification.Simplify
+    ( MonadSimplify
+    )
+import Kore.Substitute
+    ( SubstitutionVariable
+    )
+import Kore.Unification.Substitution
+    ( Substitution
+    )
+import Kore.Unification.SubstitutionNormalization
+    ( normalize
+    , normalizeSubstitution
+    )
+import qualified Kore.Unification.UnifierImpl as Unifier
+import Kore.Unification.Unify
+    ( MonadUnify
+    )
+import qualified Kore.Unification.Unify as Unifier
+import Kore.Variables.UnifiedVariable
+    ( UnifiedVariable
+    )
+
+newtype SubstitutionSimplifier simplifier =
+    SubstitutionSimplifier
+        { simplifySubstitution
+            :: forall variable
+            .  SubstitutionVariable variable
+            => Substitution variable
+            -> simplifier (OrPredicate variable)
+        }
+
+{- | A 'SubstitutionSimplifier' to use during simplification.
+
+If the 'Substitution' cannot be normalized, this simplifier moves the
+denormalized part into the predicate, but returns the normalized part as a
+substitution.
+
+ -}
+simplification :: MonadSimplify simplifier => SubstitutionSimplifier simplifier
+simplification =
+    SubstitutionSimplifier worker
+  where
+    worker substitution = maybeUnwrapSubstitution $ do
+        deduplicated <-
+            -- TODO (thomas.tuegel): If substitution de-duplication fails with a
+            -- unification error, this will still discard the entire
+            -- substitution into the predicate. Fortunately, that seems to be
+            -- rare enough to discount for now.
+            Unifier.deduplicateSubstitution substitution
```

Comment:
So, maybe this does not change existing behaviour (I didn't check that), but I have to ask this:

In simplification we use matching. I thought that if we were simplifying the matching substitution, we were supposed to also use matching, or we shouldn't simplify it at all. However, here we're calling `deduplicateSubstitution`, which, I think, is calling `simplifyAnds`, which is using unification. So, did I make a mistake in my assumptions?

Reply:
> I thought that if we were simplifying the matching substitution, we were supposed to also use matching, or we shouldn't simplify it at all.

We never simplify the substitution during matching. Matching constructs the substitution so that it is always normalized; if it's ever non-normalized, matching fails. This would be used to during predicate simplification after applying (rewrite or equality) rules, after the rule is fully applied, or on the remainders. Unification is being used to simplify `\and` patterns here, so I think it's appropriate.